File: ghostwriter/storage.py
"""
Load/save JSON profiles from profiles/ directory.
"""

import json
import os
from pathlib import Path


PROFILES_DIR = Path(__file__).parent.parent / 'profiles'


def ensure_profiles_dir():
    """Ensure profiles directory exists."""
    PROFILES_DIR.mkdir(exist_ok=True)


def list_profiles():
    """List all profile names (without .json extension)."""
    ensure_profiles_dir()
    files = [f.name[:-5] for f in PROFILES_DIR.glob('*.json')]
    return sorted(files)


def load_profile(name):
    """Load a profile by name. Returns dict or None if not found."""
    ensure_profiles_dir()
    path = PROFILES_DIR / f'{name}.json'
    if not path.exists():
        return None
    with open(path, 'r', encoding='utf-8') as f:
        return json.load(f)


def save_profile(name, fingerprint):
    """Save a fingerprint to a profile JSON file."""
    ensure_profiles_dir()
    path = PROFILES_DIR / f'{name}.json'
    with open(path, 'w', encoding='utf-8') as f:
        json.dump(fingerprint, f, indent=2)


def delete_profile(name):
    """Delete a profile."""
    ensure_profiles_dir()
    path = PROFILES_DIR / f'{name}.json'
    if path.exists():
        path.unlink()
        return True
    return False


def profile_exists(name):
    """Check if a profile exists."""
    ensure_profiles_dir()
    return (PROFILES_DIR / f'{name}.json').exists()

File: docs/js/storage.js
/**
 * localStorage wrapper for profile management.
 * Stores both the profile (result of buildProfile) and raw samples (for rebuilding).
 */

const PROFILE_PREFIX = 'ghostwriter:profile:';
const SAMPLES_PREFIX = 'ghostwriter:samples:';

/**
 * Save a profile and its samples to localStorage.
 */
export function saveProfile(name, profile, samples = []) {
  const profileKey = PROFILE_PREFIX + name;
  const samplesKey = SAMPLES_PREFIX + name;
  localStorage.setItem(profileKey, JSON.stringify(profile));
  localStorage.setItem(samplesKey, JSON.stringify(samples));
}

/**
 * Load a profile from localStorage.
 */
export function loadProfile(name) {
  const profileKey = PROFILE_PREFIX + name;
  const data = localStorage.getItem(profileKey);
  if (!data) return null;
  try {
    return JSON.parse(data);
  } catch {
    return null;
  }
}

/**
 * Load samples for a profile.
 */
export function loadSamples(name) {
  const samplesKey = SAMPLES_PREFIX + name;
  const data = localStorage.getItem(samplesKey);
  if (!data) return [];
  try {
    return JSON.parse(data);
  } catch {
    return [];
  }
}

/**
 * List all saved profile names.
 */
export function listProfiles() {
  const profiles = [];
  for (let i = 0; i < localStorage.length; i++) {
    const key = localStorage.key(i);
    if (key.startsWith(PROFILE_PREFIX)) {
      const name = key.slice(PROFILE_PREFIX.length);
      profiles.push(name);
    }
  }
  return profiles.sort();
}

/**
 * Delete a profile and its samples from localStorage.
 */
export function deleteProfile(name) {
  const profileKey = PROFILE_PREFIX + name;
  const samplesKey = SAMPLES_PREFIX + name;
  localStorage.removeItem(profileKey);
  localStorage.removeItem(samplesKey);
}

/**
 * Check if a profile exists.
 */
export function profileExists(name) {
  const profileKey = PROFILE_PREFIX + name;
  return localStorage.getItem(profileKey) !== null;
}
